### DISCUSSION TITLE
How to add "max_unpool2d" operator to onnxruntime?

### DISCUSSION THREAD

**Discussion by ArtyZe** [microsoft/onnxruntime#4610] [General]
Hello, 
       Thanks for your great job. Now I have few questions about adding ops to onnxruntime.
1.  As "max_unpool2d" is a torch op which is not able to run in onnxruntime, so I plan to add this op to onnxruntime. But I don't know how to write the source file and head file, especially compute function, as example of onnxruntime is useless now.
![image](https://user-images.githubusercontent.com/32027268/88373208-fe7dcf80-cdc9-11ea-92e1-48b10416a8e0.png)
Could you update it when you are convenient?
2. If my op domain is same as others, which should I use here?
![image](https://user-images.githubusercontent.com/32027268/88373490-7fd56200-cdca-11ea-8a20-cd154a5ab24f.png)
Thanks :+1:

**Comment by jcwchen** [comment]
Hi @ArtyZe,
1. Sorry for the mistake. The right link is as follows: 
https://github.com/onnx/tutorials/blob/master/PyTorchCustomOperator/ort_custom_op/custom_op.cc

2. Will you use some custom opsets from their domain? If so, I think you should use the same op domain as theirs. 

Thank you for reporting this document issue!

**Comment by ArtyZe** [comment]
> Hi @ArtyZe,
> 
> 1. Sorry for the mistake. The right link is as follows:
>    https://github.com/onnx/tutorials/blob/master/PyTorchCustomOperator/ort_custom_op/custom_op.cc
> 2. Will you use some custom opsets from their domain? If so, I think you should use the same op domain as theirs.
> 
> Thank you for reporting this document issue!

Thanks for the help!
As what describled in doc,
![image](https://user-images.githubusercontent.com/32027268/88510689-18abed80-d016-11ea-9331-b58dd39670d1.png)
now I compile it successfully. Got 
![image](https://user-images.githubusercontent.com/32027268/88510637-0762e100-d016-11ea-8211-9034b9ea8dfc.png)
So where should I put this lib file to support onnxruntime test model with max_unpool and other ops like conv2d? That means what should I do now to let my onnxruntime session call this function:
![image](https://user-images.githubusercontent.com/32027268/88518450-17cd8880-d023-11ea-8a67-c16c07dd9aac.png)

Actually cmake file link is also broken, so I have no idea compile my op to lib or bin file.
![image](https://user-images.githubusercontent.com/32027268/88517722-e30d0180-d021-11ea-8054-56728e3a3e60.png)

**Comment by jcwchen** [comment]
Hi @ArtyZe, 
Sorry for the late reply. The right `CmakeLists.txt` link should be: 
https://github.com/onnx/tutorials/blob/master/PyTorchCustomOperator/ort_custom_op/CMakeLists.txt
There are some urls broken as you mentioned. You can find the correct links in this PR: https://github.com/onnx/tutorials/pull/206/files

Hope it helps. Thanks.

**Reply by harshamedu** [reply]
Hi,

   I am defining a model in onnx. I want to define a custom op. How can I do in python ?
I get 'error: PyOp is not a registered function/op' for the following code.

import onnx
import onnxruntime as ort

A = onnx.helper.make_tensor_value_info('A', onnx.TensorProto.FLOAT, [4])
B = onnx.helper.make_tensor_value_info('B', onnx.TensorProto.FLOAT, [4])
C = onnx.helper.make_tensor_value_info('C', onnx.TensorProto.FLOAT, [4])
D = onnx.helper.make_tensor_value_info('D', onnx.TensorProto.FLOAT, [4])
E = onnx.helper.make_tensor_value_info('E', onnx.TensorProto.FLOAT, [4])
F = onnx.helper.make_tensor_value_info('F', onnx.TensorProto.FLOAT, [4])

ad1_node = onnx.helper.make_node('Add', ['A', 'B'], ['S'])

mul_node = onnx.helper.make_node('Mul', ['C','D'], ['P'])

ad2_node = onnx.helper.make_node('Add', ['S', 'P'], ['H'])

py1_node = onnx.helper.make_node(op_type = 'PyOp', #required, must be 'PyOp'
                            inputs = ['H'], #required
                            outputs = ['F'], #required
                            domain = 'pyopadd_2', #required, must be unique
                            input_types = [onnx.TensorProto.FLOAT], #required
                            output_types = [onnx.TensorProto.FLOAT], #required
                            module = 'mymodule', #required
                            class_name = 'Add_2', #required
                            compute = 'compute') #optional, 'compute' by default

graph = onnx.helper.make_graph([ad1_node,mul_node,ad2_node, py1_node], 'multi_pyop_graph', [A,B,C,D], [F])
model = onnx.helper.make_model(graph, opset_imports=[onnx.helper.make_opsetid('pyopadd_2', 1)], producer_name = 'pyop_model')
onnx.save(model, './modeltemp.onnx')

ort_session = ort.InferenceSession('./modeltemp.onnx')
ort_output = ort_session.run(["F"], {'A':[1,2,3,4], 'B':[1,1,1,1], 'C':[2,2,2,2], 'D':[3,3,3,3]})
print(ort_output)


mymodule.py
class Add_2:
    def compute(self, S):
        return S+2

Thanks.
